### PR TITLE
ParmParse: Math expression support

### PR DESCRIPTION
**Pull request by WeiqunZhang**
ParmParse: Math expression support

Support math expression for integers and floating point numbers in
ParmParse. For example,

    n_cell = 128
    amrex.n_cell = {n_cell*2} 8 {16**2}

becomes

    n_cell = 128
    amrex.n_cell = 256 8 256

Note that the user must use `{}` to indicate that the expression needs to be
evaluate with amrex::Parser. If the expression contains another variable, it
will be looked up by ParmParse.

ParmParse's constructor now accepts an optional second argument,
`parser_prefix`. When a variable in a math expression is being looked up, it
will first try to find it by using the exact name of the variable. If this
attempt fails and the `ParmParse` object has a non-empty `parser_prefix`, it
will try again, this time looking up the variable by prefixing its name with
`parser_prefix` followed by a `.`. For example,

    physical_constants.c = 3.e8
    amrex.pi = 3.14
    amrex.foo = { sin(amrex.pi/2) * c**2 }

will give `foo = 8.999997146e+16` if we do

    amrex::ParmParse pp("amrex", "physical_constants");
    double foo;
    pp.get("foo", foo);

We also have to make a change to amrex::Parser and iParser. The variable
names can contain `.` now.

### CODE REVIEW

**Comment by WeiqunZhang** [comment]
Depend on #4031

**Comment by BenWibking** [comment]
What is the order of evaluation of variables? Will it complain if I do:
```
code.a = {code.b}
code.b = {code.a}
```

**Comment by WeiqunZhang** [comment]
The recursion will not stop until it segfaults. @BenWibking thanks for pointing this out. We should give a warning.

As for evaluation orders, I have added an example to the documentation. Does that answer your question?

```
The variables in :cpp:`ParmParse` math expressions are not evaluated until
they are referenced. If a variable is defined multiple times, the last
occurrence will override previous ones even if it appears after the variable
has been referenced. This behavior is demonstrated in the following example.

.. highlight:: python

::

    foo.a = 1                                                                                                                                                                
    foo.b = {foo.a}                                                                                                                                                          
    foo.a = 2                                                                                                                                                                

will become

.. highlight:: python

::

    foo.a = 2                                                                                                                                                                
    foo.b = 2                 
```

**Comment by WeiqunZhang** [comment]
We just discussed this in WarpX meeting.

* @AlexanderSinn prefers `"` `"` to `{}`.
* @AlexanderSinn suggested that a static `parser_prefix` in `ParmParse` (in addition to the non-static member)
* @AlexanderSinn and @dpgrote mentioned that ideally it would be easier for the users if neither `"` nor `{}` is needed. But I think this creates ambiguities and it's actually error prone for the users because the result depends on which `ParmParse` functions (`get`, `getarr` etc.) are used. For example
    ```
    amr.n_cell = 64 64 +16 128 # four numbers or three?
    foo = sin(x) -cos(y) # is foo a scalar or two numbers sin(x) and -cos(y)?
    ```

**Comment by WeiqunZhang** [comment]
```
code.a = {code.b}
code.b = {code.a}
```
will result in
```
amrex::Error::0::ParmParse: recursive reference to {code.b} is not allowed !!!
SIGABRT
```

**Comment by AlexanderSinn** [comment]
How about if the user does not add `{}` or `""`, the expression is not allowed to contain a space (and maybe also not equals), but otherwise it works without `{}` or `""`.

**Comment by WeiqunZhang** [comment]
This is from https://github.com/Hi-PACE/hipace/blob/development/examples/laser/inputs_SI
```
my_constants.kp_inv = 10.e-6
geometry.prob_lo     = -6.*kp_inv   -6.*kp_inv   -8.*kp_inv
geometry.prob_hi     =  6.*kp_inv    6.*kp_inv    6.*kp_inv
```
They contain spaces. So for this to work, we would need to use `{}` or `""`. In Hipace++, this is handled with a special `fillWithParserArr` function. I haven't tried yet. But if we use `""` as follows, it seems that it should work with both Hipace++ and this PR (modified to use `""`).
```
my_constants.kp_inv = 10.e-6
geometry.prob_lo     = "-6.*kp_inv"   "-6.*kp_inv"   "-8.*kp_inv"
geometry.prob_hi     =  "6.*kp_inv"    "6.*kp_inv"    "6.*kp_inv"
```
On the other hand, if we use `{}`, it might be easier for you to detect in Hipace++ and try to maintain backward compatibility.

**Comment by AlexanderSinn** [comment]
With "the expression is not allowed to contain a space" I mean the individual expressions that are given to the parser, so that ParmParse can simply split the line by spaces so that
```
geometry.prob_lo     = -6.*kp_inv   -6.*kp_inv   -8.*kp_inv
geometry.prob_hi     = "-6. * kp_inv"   "-6. * kp_inv"   "-8. * kp_inv"
```
would work, but not
```
geometry.prob_lo     = -6. * kp_inv   -6. * kp_inv   -8. * kp_inv
```
The second version would currently get an error in hipace along the lines of: wrong number of arguments, expected 3 got 9.
One special thing about the current implementation in hipace is that for input parameters that are known to be a scalar value, it is not split by spaces and everything including the spaces is parsed to the Parser (more precisely it is split by spaces by ParmParse querryarr but then all elements are combined again). However, I'm fine with changing this so that scalar parameters also don't allow spaces without escaping.

**Comment by AlexanderSinn** [comment]
Note that `geometry.prob_lo` and `geometry.prob_hi` are amrex parameters, but they are parsed and replaced by the hipace functions.

**Comment by WeiqunZhang** [comment]
I am worried about it being too sensitive to empty spaces. Once the users start using math expressions, I suspect some of them might be quite long. It would then suddenly fail, if it accidentally has an unwanted space.

**Comment by dpgrote** [comment]
WarpX is doing the same thing as hipace. When a scalar is read in, it reads in the parameter as an array of strings and concatenates them and then parses the result. Then for arrays, there can be no spaces in each element of the array. This gives the very nice well defined and easy to read expressions like the example above.
```
geometry.prob_lo     = -6.*kp_inv   -6.*kp_inv   -8.*kp_inv
```
I've taken this same code and put it in MauiC as well, giving the same nice format. It would be a shame to uglify this and force the use of `{}` or `""`.

In many places in WarpX and MauiC, there are already checks on the number of elements when arrays are read in so this will catch any extraneous spaces that mighty appear anyway. To make every input parameter less nice visually just in case that rare long expression might have an extra space seems like a poor trade off.

**Comment by WeiqunZhang** [comment]
If we have
```
a = 1 -2
```
and
```
ParmParse pp;
int a;
pp.get("a", a);
```
What should `a` be? @dpgrote @AlexanderSinn  I think @dpgrote 's answer is -1, whereas @AlexanderSinn 's is 1.

**Comment by dpgrote** [comment]
I'm sure that hipace copied the code from WarpX so it would be doing the same thing and get -1. His description is the same "it is split by spaces by ParmParse querryarr but then all elements are combined again".

**Comment by WeiqunZhang** [comment]
That's not backward compatible. Currently with the development branch we get 1.

**Comment by WeiqunZhang** [comment]
If we don't use `{}` or `""`, all white spaces need to be significant for int and real unless they are inside `""`. We cannot just combine the whole line into a single scalar.

**Comment by WeiqunZhang** [comment]
I suspect that that might create more issues for WarpX and Hipace++ if both codes migrate to the new ParmParse.

**Comment by WeiqunZhang** [comment]
By the way, in all seriousness, my opinion is python would be a better language if it uses `{}`.

**Comment by dpgrote** [comment]
> That's not backward compatible. Currently with the development branch we get 1.

Ok, so for scalars parameters now it only reads in the first element of a space separate array. I would be fine keeping it this way, specifying no spaces in scalar expressions. This would be consistent with no spaces in array elements. Alex also is fine with this. This is not backwards compatible with WarpX but is probably not a problem since uses of spaces in scalars is probably quite rare anyway.

**Comment by dpgrote** [comment]
> I suspect that that might create more issues for WarpX and Hipace++ if both codes migrate to the new ParmParse.

Requiring `{}` or `""` would create far more issues. Almost every input file would need to be updated with extensive changes. With requiring no spaces in scalars, there would probably be very few updates required.

**Comment by WeiqunZhang** [comment]
WarpX has things like `my_constants.n0 = beam_charge / (q_e * sigmax * sigmay * sigmaz * (2.*pi)**(3./2.))`. Maybe we can add functions like `ParmParse::queryWithParser` that will combine everything into one string before parsing.

**Comment by AlexanderSinn** [comment]
> If we have
> 
> ```
> a = 1 -2
> ```
> 
> and
> 
> ```
> ParmParse pp;
> int a;
> pp.get("a", a);
> ```
> 
> What should `a` be? @dpgrote @AlexanderSinn I think @dpgrote 's answer is -1, whereas @AlexanderSinn 's is 1.

To me this should either be  `-1` or error: expected one element got two.

**Comment by AlexanderSinn** [comment]
The behavior where not everything is read in (which would result in 1) is intentionally avoided and asserted in hipace for all parameters.

**Comment by WeiqunZhang** [comment]
https://github.com/AMReX-Codes/amrex/blob/bc86eacae9e7204ba1689fd9f05af4e5e62529af/Docs/sphinx_documentation/source/Basics.rst#math-expressions

**Comment by WeiqunZhang** [comment]
This PR also depends on #4038.

**Comment by WeiqunZhang** [comment]
As for relationship with #4035, either this one or #4035 can be merged after #4031. The two are likely to have conflicts, but I will fix them after one is merged.

File: Docs/sphinx_documentation/source/Basics.rst
Comment (dpgrote) [inline Docs/sphinx_documentation/source/Basics.rst]: Typo `queryWithParser`

Comment (WeiqunZhang) [inline Docs/sphinx_documentation/source/Basics.rst]:
```suggestion
:cpp:`ParmParse::queryWithParser` or :cpp:`ParmParse::getWithParser`. If the
```

File: Docs/sphinx_documentation/source/Basics.rst
Comment (dpgrote) [inline Docs/sphinx_documentation/source/Basics.rst:392]: Interesting. Is there a reason why you have two different parser prefix options?

Comment (WeiqunZhang) [reply Docs/sphinx_documentation/source/Basics.rst:392]: Initially there was only one that was stored as a non-static member. It's safer, but it would require passing the prefix string to the constructor. So a static member was added to give people more options.

**Comment by WeiqunZhang** [comment]
Note that allowing spaces inside `""` comes for free because that's how we handle strings.

**Review by dpgrote — APPROVED** [review]
Looks good!

**Comment by AlexanderSinn** [comment]
I tested this PR by using it in HiPACE++ and replacing the custom parser functions with amrex pp.get and pp.query. It did mostly work with existing input scripts. I had to add a RealVect overload. I didn't use queryWithParser and there actually were silent errors in the script where "" would be needed (as expected). To make parser functions that are evaluated on the GPU, it would be nice if some of the variable reading functionality from pp_parser could be reused by users. One issue I found was that for `beam1.num_particles = 5e9` which is queried as Long, the iParser couldn't parse `5e9` and after I changed it to `5*10^9` it overflowed to 705032704 without giving an error.

**Comment by WeiqunZhang** [comment]
We could change iParser to use long int. As for `5e9`, we could support that if that does not cause any confusion. We can do that in a separate PR.

Could you elaborate on "To make parser functions that are evaluated on the GPU, it would be nice if some of the variable reading functionality from pp_parser could be reused by users."?

**Comment by AlexanderSinn** [comment]
We use amrex::Parser and amrex::ParserExecutor to evaluate functions on the GPU that come from the input script. For this we currently get them as a string from ParmParse, give that to the Parser and then replace all the unknown symbols with constants from ParmParse, just like it's done in pp_parser.

**Comment by WeiqunZhang** [comment]
Okay, I think I get what you mean. We can add a `Parser ParmParse::makeParser (const char* name)`, where `name` is a ParmParse name like `f(x,y)` in `f(x,y) = a*x+b*y`, or we we add a `Parser ParmParse::makeParser (std::string const& function_body)`, where `function_body` is `a*x+b*y`. In both cases, we will set known constants. Or should we implement both versions? How to avoid confusion?

Or we can take the hipace++ approach. https://github.com/Hi-PACE/hipace/blob/5d8dda3cad53b5fc051eaa98937ba3eb8d3c9bb0/src/utils/Parser.H#L510

**Comment by WeiqunZhang** [comment]
As for iparser, maybe it's better to support `5'000'000'000` than `5e9`.

**Comment by AlexanderSinn** [comment]
I prefer the `std::string const& function_body` version as its more generall and can be combined with get/query/getarr/queryarr and other sources as needed.

**Comment by AlexanderSinn** [comment]
For iParser `5'000'000'000` would be nice, but I think also `5e9` makes sense so long as there is an assert that the number is an integer and fits inside int/long (`4.32e1`and `5e20` would not be allowed). `amrex.the_arena_init_size` should also benefit from this.

**Comment by WeiqunZhang** [comment]
Added `makeParser` and `makeIParser`. I will open another PR for switching to long long int in IParser.

File: Src/Base/AMReX_ParmParse.H
Comment (WeiqunZhang) [inline Src/Base/AMReX_ParmParse.H]:
```suggestion
    [[nodiscard]] Parser makeParser (std::string const& func,
                                     Vector<std::string> const& vars) const;
```

File: Src/Base/AMReX_ParmParse.H
Comment (WeiqunZhang) [inline Src/Base/AMReX_ParmParse.H]:
```suggestion
    [[nodiscard]] IParser makeIParser (std::string const& func,
                                       Vector<std::string> const& vars) const;
```

**Review by atmyers — APPROVED** [review]

**Pull request merged**
